### PR TITLE
refactoring generate.go queries

### PR DESCRIPTION
**Pull request by diaabadr**
### Description

Enhanced query performance in the `generate.go` by refactoring the insertion process. Replaced individual insertions within a loop with a single insert statement.

### Changes

- Refactored all insertion and update queries to be in single SQL statement
- Added function to utils to extract the values of each statement.

### Related Issues

#358

### Checklist

- [X] Tests included
- [X] Build pass
- [X] Documentation
- [X] Code format and docstring

### Execution time comparison based on the provided constant values in the file:
- **Before:** 84.931081115 seconds
- **After:** 0.993061923 seconds

### CODE REVIEW

File: grid-proxy/tools/db/utils.go [hunk 1]
```diff
@@ -44,6 +45,15 @@ func max(a, b uint64) uint64 {
 	return b
 }
 
+func extractValuesFromInsertQuery(q string) (string, error) {
```
Comment (xmonader) [inline grid-proxy/tools/db/utils.go]: Add test for this function and include docs for the behavior

File: grid-proxy/tools/db/generate.go [hunk 2]
```diff
@@ -70,14 +83,32 @@ func generateTwins(db *sql.DB) error {
 			twin_id:      i,
 			grid_version: 3,
 		}
-		if _, err := db.Exec(insertQuery(&twin)); err != nil {
+
+		value, err := extractValuesFromInsertQuery(insertQuery(&twin))
+		if err != nil {
+			panic(err)
```
Comment (xmonader) [inline grid-proxy/tools/db/generate.go]: the function returns the error, and the caller in generateData is acting on that value, so don't panic here, please refactor to return and the call can do whatever.

File: grid-proxy/tools/db/generate.go [hunk 3]
```diff
@@ -361,34 +534,82 @@ func generateNodes(db *sql.DB) error {
...
+			publicConfigValues = append(publicConfigValues)
+
+		}
+	}
+
+	if len(locationValues) != 0 {
+		query := "INSERT INTO location (id, longitude, latitude) VALUES"
+		query += strings.Join(locationValues, ",") + ";"
+		if _, err := db.Exec(query); err != nil {
+			panic(err)
```
Comment (xmonader) [inline grid-proxy/tools/db/generate.go]: same comments as in generateTwins, please apply everywhere.

File: grid-proxy/tools/db/generate.go [hunk 4]
```diff
@@ -70,14 +83,32 @@ func generateTwins(db *sql.DB) error {
 			twin_id:      i,
 			grid_version: 3,
 		}
-		if _, err := db.Exec(insertQuery(&twin)); err != nil {
+
+		value, err := extractValuesFromInsertQuery(insertQuery(&twin))
```
Comment (mariobassem) [inline grid-proxy/tools/db/generate.go]: i see that insertQuery is only called in extractValuesFromInsertQuery, hence i think it should be called internally if it's needed at all.

File: grid-proxy/tools/db/generate.go [hunk 5]
```diff
@@ -70,14 +83,32 @@ func generateTwins(db *sql.DB) error {
...
+		if err != nil {
+			panic(err)
+		}
+		values = append(values, value)
+	}
+
+	if len(values) != 0 {
```
Comment (mariobassem) [inline grid-proxy/tools/db/generate.go]: for better readability, wrap this block in a function whose only purpose is to insert twin records

File: grid-proxy/tools/db/generate.go [hunk 6]
```diff
@@ -91,17 +122,39 @@ func generatePublicIPs(db *sql.DB) error {
...
 			panic(err)
 		}
 	}
+	fmt.Println("public IPs generated")
+
 	return nil
 }
 
 func generateFarms(db *sql.DB) error {
+	var values []string
```
Comment (mariobassem) [inline grid-proxy/tools/db/generate.go]: rename to something more descriptive

File: grid-proxy/tools/db/generate.go [hunk 7]
```diff
@@ -117,14 +170,32 @@ func generateFarms(db *sql.DB) error {
...
+		if _, err := db.Exec(query); err != nil {
+			panic(err)
+		}
+	}
+	fmt.Println("Farms generated")
+
 	return nil
 }
 
 func generateContracts(db *sql.DB) error {
```
Comment (mariobassem) [inline grid-proxy/tools/db/generate.go]: generate contracts should generate all contracts (node, name, rent). maybe embed the three contract generation functions in it.

**Review by mariobassem — CHANGES_REQUESTED** [review]
great work Diaa. just some small comments

File: grid-proxy/tools/db/types.go [hunk 8]
```diff
@@ -54,6 +54,8 @@ type twin struct {
 	relay        string
 	public_key   string
 }
+
```
Comment (mariobassem) [inline grid-proxy/tools/db/types.go:57]: remove new line

File: grid-proxy/tools/db/utils.go [hunk 9]
```diff
@@ -2,21 +2,32 @@ package main
...
 	"fmt"
 	"math/rand"
 	"net"
 	"reflect"
 	"strings"
 )
 
-func rnd(min, max uint64) uint64 {
-	return rand.Uint64()%(max-min+1) + min
+var errMaxMinInvalid = errors.New("min cannot be greater than max")
```
Comment (mariobassem) [inline grid-proxy/tools/db/utils.go]: we don't really need to create an error variable, right?
if it's to easily check it when testing, just checking if there is an error is usually enough.

File: grid-proxy/tools/db/utils_test.go [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+	assert.Equal(t, errMaxMinInvalid, got)
+}
+
+func TestMax(t *testing.T) {
+	want := uint64(5)
+	got := max(4, 5)
+	assert.Equal(t, want, got)
+}
+
+func TestObjectToTupleString(t *testing.T) {
```
Comment (mariobassem) [inline grid-proxy/tools/db/utils_test.go:28]: great!
maybe just add one more test run in this function to test NULL values?

File: grid-proxy/tools/db/generate.go [hunk 11]
```diff
@@ -31,10 +31,10 @@ const (
 	nodeUpRatio          = .5
 	nodeCount            = 1000
 	farmCount            = 100
-	normalUsers          = 2000
+	normalUsers          = 200
```
Comment (mariobassem) [inline grid-proxy/tools/db/generate.go]: why did normal users count decrease to 200??

File: grid-proxy/tools/db/generate.go [hunk 12]
```diff
@@ -31,10 +31,10 @@ const (
 	nodeUpRatio          = .5
 	nodeCount            = 1000
```
Comment (mariobassem) [inline grid-proxy/tools/db/generate.go]: maybe increase the numbers a bit??
let's do nodes 100000, farms 10000, nodeContracts 100000, rentContracts 1000, nameContracts 1000, normalUsers 50000. 
what do you think?

File: grid-proxy/tools/db/generate.go [hunk 13]
```diff
@@ -397,37 +606,113 @@ func generateNodeGPUs(db *sql.DB) error {
...
+	fmt.Println("node GPUs generated")
+
+	return nil
+
+}
+
+func generateContracts(db *sql.DB) error {
+	billCount := 1
+	// Note: it's order senstive
+	if err := generateRentContracts(db, &billCount); err != nil {
```
Comment (mariobassem) [inline grid-proxy/tools/db/generate.go]: having to pass the billCount by reference works in our use case, but in general it's not a really good idea unless it's needed, as a pointer creates the possibility for the value to not even be there.
i believe we can work around that, right?

File: grid-proxy/tools/db/generate.go [hunk 14]
```diff
@@ -397,37 +606,113 @@ func generateNodeGPUs(db *sql.DB) error {
...
+
+	fmt.Println("node GPUs generated")
+
+	return nil
+
+}
+
+func generateContracts(db *sql.DB) error {
+	billCount := 1
+	// Note: it's order senstive
```
Comment (mariobassem) [inline grid-proxy/tools/db/generate.go]: is there a way to make it not order sensitive?

File: grid-proxy/tools/db/generate.go [hunk 15]
```diff
@@ -70,38 +82,66 @@ func generateTwins(db *sql.DB) error {
...
+		return err
 	}
+	fmt.Println("twins generated")
+
 	return nil
 }
 
 func generatePublicIPs(db *sql.DB) error {
+	var publicIPs []string
+	var nodeContracts []interface{}
```
Comment (mariobassem) [inline grid-proxy/tools/db/generate.go]: does nodeContracts need to be []interface{} ?

File: grid-proxy/tools/db/generate.go [hunk 16]
```diff
@@ -70,38 +82,66 @@ func generateTwins(db *sql.DB) error {
...
-			contract_id = createdNodeContracts[rnd(0, uint64(len(createdNodeContracts))-1)]
+			idx, err := rnd(0, uint64(len(createdNodeContracts))-1)
+			if err != nil {
+				return err
+			}
+			contract_id = createdNodeContracts[idx]
 		}
 		ip := randomIPv4()
+		farmID, err := rnd(1, farmCount)
+
```
Comment (mariobassem) [inline grid-proxy/tools/db/generate.go]: remove empty line?

File: grid-proxy/tools/db/generate.go [hunk 17]
```diff
@@ -216,47 +324,74 @@ func generateNameContracts(db *sql.DB) error {
...
+	if err := insertTuples(db, contract_bill_report{}, billReports); err != nil {
+		return err, billsStartID, contractsStartID
+	}
+
+	fmt.Println("name contracts generated")
+
+	return nil, billsStartID, contractsStartID
 }
-func generateRentContracts(db *sql.DB) error {
+func generateRentContracts(db *sql.DB, billsStartID, contractsStartID int) (error, int, int) {
```
Comment (mariobassem) [inline grid-proxy/tools/db/generate.go]: nothing mandates that an error should have a specific position an a function's returned values, but if we care about our code's readability we would stick with the convention of having the error as the last returned value, right?

File: grid-proxy/tools/db/generate.go [hunk 18]
```diff
@@ -397,37 +609,117 @@ func generateNodeGPUs(db *sql.DB) error {
...
+
+	return nil
+
+}
+
+func generateContracts(db *sql.DB) error {
+	billsStartID := 1
+	contractsStartID := 1
+
+	err, billsStartID, contractCount := generateRentContracts(db, billsStartID, contractsStartID)
```
Comment (mariobassem) [inline grid-proxy/tools/db/generate.go]: this is better from last version. 
i would go further with this and make the generateRentContract, generateNodeContracts, and generateNameContracts functions just create the string arrays of the contracts and the bill reports, and do the actual insert calls here in the generateContracts function.
but this is just nitpicking, feel free to ignore.

**Review by mariobassem — CHANGES_REQUESTED** [review]

File: grid-proxy/tools/db/generate.go (hunk 16 above)
Comment (diaabadr) [reply grid-proxy/tools/db/generate.go]: formatter does not remove them.

File: grid-proxy/tools/db/generate.go (hunk 18 above)
Comment (diaabadr) [reply grid-proxy/tools/db/generate.go]: what about generating all contracts in the functions and collecting reports from each one and inserting them once? that's the one which will make the performance better as inserting all bills at once.

File: grid-proxy/tools/db/generate.go [hunk 19]
```diff
@@ -267,72 +399,131 @@ func generateRentContracts(db *sql.DB) error {
...
+
 		up := flip(nodeUpRatio)
-		updatedAt := time.Now().Unix() - int64(rnd(60*40*3, 60*60*24*30*12))
+		periodFromLatestUpdate, err := rnd(60*40*3, 60*60*24*30*12)
+		if err != nil {
+			return err
+		}
+		updatedAt := time.Now().Unix() - int64(periodFromLatestUpdate)
+
+		periodFromLatestUpdate, err = rnd(0, 60*40*1)
```
Comment (mariobassem) [inline grid-proxy/tools/db/generate.go]: should this be in the `if up {}` block ??

**Review by mariobassem — APPROVED** [review]

**Review by mariobassem — APPROVED** [review]

**Review by xmonader — APPROVED** [review]
